### PR TITLE
Read Parquet files faster

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reading files in Parquet format is now much faster. IO and decoding are parallelized (controlled by `max_threads` setting), and only required data ranges are read.

This is still not very efficient. Possible future improvements:
 * Implement parallel reading for StorageFile (likely by implementing factory versions of MMapReadBufferFromFileDescriptor and ReadBufferFromFileDescriptorPRead). With this PR, decoding happens in max_threads threads, but they all read from one ReadBuffer, locking a mutex. No good, but still faster than one thread.
 * Do parallel reading separately from parallel parsing, to allow having different number of threads for it. In particular, lots of download threads when reading from a different region, as Parquet often wants to read lots of short ranges.
 * Do decoding ourselves instead of using arrow, to avoid some copying. Maybe reimplement everything, or maybe reuse arrow's metadata decoding and lower-level data decoding (low enough level that we don't have to copy unnecessarily, if such level exists).

### CODE REVIEW

File: src/Core/BackgroundSchedulePool.cpp [hunk 1]
```diff
@@ -157,10 +160,26 @@ BackgroundSchedulePool::BackgroundSchedulePool(size_t size_, CurrentMetrics::Met
...
+        LOG_FATAL(
+            &Poco::Logger::get("BackgroundSchedulePool/" + thread_name),
+            "Couldn't get {} threads from global thread pool: {}",
+            size_,
+            getCurrentExceptionCode() == DB::ErrorCodes::CANNOT_SCHEDULE_TASK
+                ? "Not enough threads. Please make sure max_thread_pool_size is considerably "
+                  "bigger than background_schedule_pool_size."
+                : getCurrentExceptionMessage(/* with_stacktrace */ true));
+        abort();
+    }
```
Comment (al13n321) [inline src/Core/BackgroundSchedulePool.cpp:182]: (This is unrelated to the rest of the PR. Without this, the server gives you a quiet SIGABRT if you set max_thread_pool_size too small. Now it'll log an error.)

File: src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp [hunk 2]
```diff
@@ -146,12 +146,7 @@ void AsynchronousReadIndirectBufferFromRemoteFS::prefetch(int64_t priority)
 
 void AsynchronousReadIndirectBufferFromRemoteFS::setReadUntilPosition(size_t position)
 {
-    /// Do not reinitialize internal state in case the new end of range is already included.
-    /// Actually it is likely that we will anyway reinitialize it as seek method is called after
-    /// changing end position, but seek avoiding feature might help to avoid reinitialization,
-    /// so this check is useful to save the prefetch for the time when we try to avoid seek by
-    /// reading and ignoring some data.
-    if (!read_until_position || position > *read_until_position)
+    if (!read_until_position || position != *read_until_position)
```
Comment (al13n321) [inline src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:149]: This was making read_until_position useless on Parquet because it first reads metadata at the end, then everything else. Hope this check wasn't important in some other situation.

**Comment by al13n321** [comment]
Some speed numbers. Better than before, worse than what is possible.
* Querying one column from a local file didn't get faster (or slower), it's still reading sequentially. (`SELECT AdvEngineID, COUNT(*) AS c FROM file('ds/hits.parquet') GROUP BY AdvEngineID ORDER BY c DESC LIMIT 10` 0.75 s)
* All columns, local file: ~8x faster, it's decoding in parallel (after reading sequentially). (`SELECT sum(ignore(*)) AS c FROM file('ds/hits.parquet')` 102 s -> 12.7 s)
* One column, S3, local region: 50x faster, it's both reading and decoding in parallel. (`SELECT AdvEngineID, COUNT(*) AS c FROM url('https://clickhouse-datasets-us-west-2.s3.amazonaws.com/hits.parquet') GROUP BY AdvEngineID ORDER BY c DESC LIMIT 10` 49 s -> 1 s)
* All columns, S3, local region: 4x faster. (`SELECT sum(ignore(*)) AS c FROM url('https://clickhouse-datasets-us-west-2.s3.amazonaws.com/hits.parquet')` 193 s -> 46 s)
* One column, remote region (US <- Europe): 27x faster (`SELECT AdvEngineID, COUNT(*) AS c FROM url('https://clickhouse-public-datasets.s3.amazonaws.com/hits_compatible/hits.parquet') GROUP BY AdvEngineID ORDER BY c DESC LIMIT 10` 385 s -> 14 s)
* Increasing max_threads from 16 to 128 makes the previous query another 3x faster (5 s). Increasing to 256 (more than the number of HTTP requests the query does) does nothing, still 5x slower than local region, not sure why.
* All columns, remote region: 7x faster (`SELECT sum(ignore(*)) AS c FROM url('https://clickhouse-public-datasets.s3.amazonaws.com/hits_compatible/hits.parquet')` 800 s -> 115 s)
* All of the above gets about the same speed between `url()` and `s3()` table functions. (As it should, S3 is just HTTP with extra steps.) (Before this PR, this wasn't always the case; the comparisons above are using `url()`.)
* If the query does anything nontrivial with the fetched data, it gets bottlenecked on https://github.com/ClickHouse/ClickHouse/issues/38755 instead.

**Comment by Avogar** [comment]
I will start reviewing the changes. Looks very promising, great work!
Also please check tests failures, most of them are related.

File: src/IO/ParallelReadBuffer.cpp [hunk 3]
```diff
@@ -121,26 +137,25 @@ off_t ParallelReadBuffer::seek(off_t offset, int whence)
...
                 working_buffer = internal_buffer = Buffer(current_segment.data(), current_segment.data() + current_segment.size());
-                current_position += current_segment.size();
                 pos = working_buffer.end() - (current_position - offset);
                 addReaders();
                 return offset;
             }
-
-            current_position += next_segment.size();
         }
+        chassert(false);
```
Comment (Avogar) [inline src/IO/ParallelReadBuffer.cpp:150]: Why?

**Comment by Avogar** [comment]
Looks good in general. I will go deep into details later.
Also, it would be really good to try to refactor/simplify code in `FormatFactory::getInputImpl` and in `ParquetBlockInputFormat`, because it was a bit hard to read it and understand all logic.

File: src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp (hunk 2 above)
Comment (kssenii) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:149]: changing this will make this case worse https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp#L230
e.g. there could be already prefetched data or data on the fly, but we will discard it for the first half of the tasks.
Though this case is not the most important, probably could be removed.

File: src/IO/ParallelReadBuffer.cpp (hunk 3 above)
Comment (al13n321) [reply src/IO/ParallelReadBuffer.cpp:150]: Just to make it obvious that this branch always returns, when skimming the code. But looks like it's more confusing than helpful, removing.

**Comment by al13n321** [comment]
> refactor/simplify code in FormatFactory::getInputImpl

Couldn't actually remove any logic from FormatFactory, each bit seems useful, almost all of it existed already, in different places. Moved things around a little to hopefully make it more clear. Lmk if you have better ideas. Especially for how to organize the whole thing better, I don't like how awkward the whole `random_access_input_creator` thing turned out, but I couldn't think of anything better that would work.

> ParquetBlockInputFormat

Same, couldn't think of a simpler way to do it (especially if we're going to make it decode columns in parallel too). Reorganized a little and added comments, open to suggestions.

File: src/IO/ReadWriteBufferFromHTTP.h [hunk 4]
```diff
@@ -743,19 +790,46 @@ namespace detail
...
-            return timegm(&info);
+                if (response.getStatus() == Poco::Net::HTTPResponse::HTTPStatus::HTTP_PARTIAL_CONTENT)
+                {
+                    *res.file_size += requested_range_begin;
+                    res.seekable = true;
+                }
+                else
+                {
+                    res.seekable = response.has("Accept-Ranges") && response.get("Accept-Ranges") == "bytes";
+                }
```
Comment (al13n321) [inline src/IO/ReadWriteBufferFromHTTP.h:820]: This changes behavior slightly: previously StorageURL would send a HEAD request with "Range: 0-" for this check, now it's without "Range" (except if this parseFileInfo() is called from nextImpl(), which is not the important call site). I expect it's ok because HTTP servers are supposed to report "Accept-Ranges: bytes" anyway (?), but I don't actually know what HTTP servers are out there and what headers they send. For S3 this works, for althttpd this didn't work even before the change, didn't try anything else.

**Comment by al13n321** [comment]
Still looking at the test failures.

**Comment by al13n321** [comment]
Ok, fixed, there were two pre-existing bugs in ReadBufferFromS3::{seek,setReadUntilPosition} that I didn't notice, which previously weren't being hit because these methods were mostly unused.

File: src/IO/ReadBufferFromS3.cpp [hunk 5]
```diff
@@ -197,15 +199,15 @@ bool ReadBufferFromS3::nextImpl()
 
 off_t ReadBufferFromS3::seek(off_t offset_, int whence)
 {
-    if (offset_ == offset && whence == SEEK_SET)
```
Comment (al13n321) [inline src/IO/ReadBufferFromS3.cpp:200]: One bug.

File: src/IO/ReadBufferFromS3.cpp [hunk 6]
```diff
@@ -269,14 +272,43 @@ void ReadBufferFromS3::setReadUntilPosition(size_t position)
...
+        if (impl)
+        {
+            if (!atEndOfRequestedRangeGuess())
+                ProfileEvents::increment(ProfileEvents::ReadBufferSeekCancelConnection);
+            offset = getPosition();
+            resetWorkingBuffer();
+            impl.reset();
+        }
         read_until_position = position;
-        impl.reset();
```
Comment (al13n321) [inline src/IO/ReadBufferFromS3.cpp:273]: Two bug (`impl` was being destroyed without resetting `working_buffer`, which points into `impl`).

File: src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp (hunk 2 above)
Comment (al13n321) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:149]: I couldn't figure out how MergeTreePrefetchedReadPool task stealing is related to setReadUntilPosition(). AFAICT:
 * Each MergeTreeReadTask gets its own IMergeTreeReader, they're not shared or reused.
 * Each MergeTreeReaderWide (IMergeTreeReader) creates its own instances of MergeTreeReaderStream.
 * Each MergeTreeReaderStream creates its own ReadBuffer.
 * So ReadBuffers are not reused across MergeTreeReadTask, so it doesn't matter which task runs on which thread, as far as setReadUntilPosition() is concerned.

Which of these steps is incorrect?

Comment (kssenii) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:149]: ReadBuffer are reused only in case the thread stole a task which reads from a part which this thread last read https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeThreadSelectProcessor.cpp#L58. 
But in the case I mentioned it does not matter if the part name is the same.
Prefetched read pool created a reader with async prefetch and it lies in MergeTreeReadTask as std::future. In the lines of code I attached, there is a tricky case when we want to steal half of the task of one thread. So the prefetch covers the first part of the task, we steal second part of the task. So then if we see in setReadUntilPosition that the right bound of the read task changed (it will be, as we split the task in two parts), we will reset the prefetch because it is always done when right bound changes.
Another point is that we could not reset prefetch in this case if we know that prefetch future is already valid and it will make this case better, currently it resets it unconditionally.

**Comment by al13n321** [comment]
Tests still fail, at least some of them because order of rows is not preserved by default anymore.

Should we make `input_format_parquet_preserve_order` setting default to true, in case anyone relies on the ordering? It's tempting to default to false to get better speed out of the box, especially in benchmarks.

**Comment by alexey-milovidov** [comment]
Set it to true by default in this pull request. Then we will merge. Then we will change it to false in another pull request. It will allow us to highlight it better in the changelog.

**Comment by al13n321** [comment]
Would be nice to add a virtual column for row index, then detect `ORDER BY _row_idx` to switch to order-preserving mode. Doesn't help with breaking people's existing queries, just a prettier interface than a setting.

File: src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp [hunk 7]
```diff
@@ -34,108 +41,329 @@ namespace ErrorCodes
...
+        }
+
+        initializeRowGroupReader(row_group_idx);
     }
-    else
+
+
+    auto batch = row_group.record_batch_reader->Next();
+    if (!batch.ok())
+        throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA, "Error while reading Parquet data: {}", batch.status().ToString());
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp:244]: This code is executed with unlocked `lock` and above we have `SCOPE_EXIT({ chassert(lock.owns_lock()); });` and assert will fail. I think we can call `end_of_row_group()` before throwing an exception.

File: src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp [hunk 8]
```diff
@@ -34,108 +41,329 @@ namespace ErrorCodes
...
 
-    file_reader.reset();
-    current_record_batch_reader.reset();
-    column_indices.clear();
-    row_group_current = 0;
-    block_missing_values.clear();
+    /// If defaults_for_omitted_fields is true, calculate the default values from default expression for omitted fields.
+    /// Otherwise fill the missing columns with zero values of its type.
+    BlockMissingValues * block_missing_values_ptr = format_settings.defaults_for_omitted_fields ? &res.block_missing_values : nullptr;
+    row_group.arrow_column_to_ch_column->arrowTableToCHChunk(res.chunk, *tmp_table, (*tmp_table)->num_rows(), block_missing_values_ptr);
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp:259]: Similare to above, this can throw an exception and we will have failed assert

File: src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp (hunk 2 above)
Comment (kssenii) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp]: > MergeTreeReadTask is never split into multiple tasks.  Its set of ranges is never modified. Work stealing moves tasks around but doesn't mutate/split/merge/create/destroy them.

It is not true in the case I was talking about, it was on the link of my first comment https://github.com/ClickHouse/ClickHouse/blob/35572321a14d617cfd110a48d8d3416615bd75c9/src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp#L230

> So, moving tasks between MergeTreePrefetchedReadPool::threads_tasks shouldn't affect setReadUntilPosition() calls much?

it affects because a reader with a prefetch can be created before the task is taken by some thread in getNewTask, and it could have a position already set to the boundaries of the task before it was split and some data according to initial task range was requested (though it is only data of 1 prefetch buffer size starting from the start of task range):

https://github.com/ClickHouse/ClickHouse/blob/35572321a14d617cfd110a48d8d3416615bd75c9/src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp#L106

so we have this case 

1. we created a prefetched task with [10, 30]
2. created a reader
3. decided to make a prefetch
4. seeked to 10
5. called setReadUntilPosition to 30
6. async prefetch is somewhere on the fly
7. then actual reading starts
8. some thread comes in getNewTask and decides to steal
9. the only option to steal is that prefetched task
10. he gets the second half of the task while the first half of the task remains, this first half will have a future with prefetched reader
11. when the next thread gets this the first half of splited task, it also does setReadUntilPosition on the reader it got from std::future from inside the task
12. now this setReadUntilPosition will be 20, though was 30, so it will reset the prefetch with the current changes


> I think I'll give up trying to understand how exactly merge tree reading works and just add that check for prefetch_future.valid()

yep, also an option.
Also as I said, this case with splitting the task is not an important case (I do not remember if there was perf difference from it), we can remove it as well from the code of prefetched read pool.

Comment (al13n321) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp]: > he gets the second half of the task while the first half of the task remains, this first half will have a future with prefetched reader

Where's the code that does this? The code after `/// Get second half of the tasks` only moves whole `MergeTreeReadTaskPtr`s between two `deque`s, doesn't mutate or create `MergeTreeReadTask`s.

Comment (kssenii) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp]: ah... indeed, I was thinking about the case which is not implemented, sorry for confusing you...

**Review by kssenii — COMMENTED** [review]
> Couldn't reproduce the CachedOnDiskReadBufferFromFile assertion failure locally: https://s3.amazonaws.com/clickhouse-test-reports/47964/8dc28ad500c79074bb7d638b5ec5f06b3c6ed30e/stress_test__debug_.html . 

I've never seen this assertion fail before, I'd think it is related to changes

> another CachedOnDiskReadBufferFromFile wants to read the same segment. It takes the remote_file_reader, probably seeks it to 0 (within working_buffer), then fails an assert when getFileOffsetOfBufferEnd() (732) is different from downloaded_size (0).

I'd think the the problem is in another place, because see these lines of logs
```
2023.04.07 07:44:39.755076 [ 3363 ] {befe85ec-9b42-4d3c-b454-c75455a15a04} <Test> CachedOnDiskReadBufferFromFile(00170_test/yaq/fqcczxfqquhmvacyzmwqwnoglwagg): Read 732 bytes, read type REMOTE_FS_READ_AND_PUT_IN_CACHE, position: 0, offset: 732
2023.04.07 07:44:39.777827 [ 3363 ] {befe85ec-9b42-4d3c-b454-c75455a15a04} <Test> FileSegment(b2c2cee274e671de0e4265cbb005a440) : [0, 731]: Updated state from DOWNLOADING to PARTIALLY DOWNLOADED
2023.04.07 07:44:39.777959 [ 3363 ] {befe85ec-9b42-4d3c-b454-c75455a15a04} <Test> FileSegment(b2c2cee274e671de0e4265cbb005a440) : [0, 731]: Resetting downloader from befe85ec-9b42-4d3c-b454-c75455a15a04:3363
2023.04.07 07:44:39.778193 [ 3363 ] {befe85ec-9b42-4d3c-b454-c75455a15a04} <Test> FileSegment(b2c2cee274e671de0e4265cbb005a440) : [0, 731]: Complete batch. (File segment: [0, 731], key: b2c2cee274e671de0e4265cbb005a440, state: PARTIALLY_DOWNLOADED, downloaded size: 0, reserved size: 732, downloader id: None, current write offset: 0, first non-downloaded offset: 0, caller id: befe85ec-9b42-4d3c-b454-c75455a15a04:3363, detached: 0, kind: Regular, unbound: 0)
```

732 bytes were read, then we returned from `nextImpl` without writing anything to cache and there is no error in log,  this is not normal. Probably the reason is that we passed `working_buffer.begin()` instead of `position()` (therefore it could silently write nothing, let's add an assertion to `FileSegment::write` like `assertdata != nullptr)`?) but this is a bit strange, there was a guarantee that impl buffer is not seekable (the `restricted_seek` check guaranteed that) and we always passed external buffer forward into `impl`, so it was fine to use `working_buffer.begin()`.

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 9]
```diff
@@ -835,8 +828,6 @@ bool CachedOnDiskReadBufferFromFile::nextImplStep()
 
     chassert(!internal_buffer.empty());
 
-    swap(*implementation_buffer);
```
[the remark this answers is not in the document]
Comment (al13n321) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:836]: Ah, I see, `use_external_buffer` in ReadBufferFromS3 is enabled, and it means that we have to assign its `internal_buffer`.

Would it still be better to `implementation_buffer->set(...)` (like `ReadBufferFromS3` does with its `internal_buffer`) instead of `swap()`, to worry less about missing a matching `swap()` later (on exception or early return)?

The other `swap()` wasn't in SCOPE_EXIT; wouldn't that leave the CachedOnDiskReadBufferFromFile in invalid state, with internal_buffer empty? I guess that's fine because we don't expect the CachedOnDiskReadBufferFromFile to be used after its `nextImpl()` throws?

Is this a general convention for all ReadBuffers - that they shouldn't be used after nextImple() threw? This doesn't seem to hold for `FileSegment::remote_file_reader` - it can be reused by a later CachedOnDiskReadBufferFromFile even if the ReadBuffer threw an exception into an earlier CachedOnDiskReadBufferFromFile; right? Would it make sense to clarify the contract for what can be done to ReadBuffers after they threw, and add a comment in ReadBuffer? Or it it deliberately underspecified?

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 10]
```diff
@@ -535,7 +536,7 @@ void CachedOnDiskReadBufferFromFile::predownload(FileSegmentPtr & file_segment)
         /// download from offset a'' < a', but return buffer from offset a'.
         LOG_TEST(log, "Bytes to predownload: {}, caller_id: {}", bytes_to_predownload, FileSegment::getCallerId());
 
-        chassert(implementation_buffer->getFileOffsetOfBufferEnd() == file_segment->getCurrentWriteOffset());
+        chassert(static_cast<size_t>(implementation_buffer->getPosition()) == file_segment->getCurrentWriteOffset());
```
[the remark this answers is not in the document]
Comment (al13n321) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:539]: The idea was to make CachedOnDiskReadBufferFromFile less picky to `implementation_buffers`'s state and implementation. Currently CachedOnDiskReadBufferFromFile requires that:
 * `implementation_buffer->nextImpl()` doesn't assign `nextimpl_working_buffer_offset`, i.e. that `implementation_buffer->position() == implementation_buffer->buffer().begin()` here.
 * When reusing a pre-existing ReadBuffer from the FileSegment, that ReadBuffer's position() must be at the end of its working_buffer, and its `getFileOffsetOfBufferEnd()` must be at `file_segment->getCurrentWriteOffset()`. This assumption is what was breaking here.

This commit removes both requirements. We just take a ReadBuffer, seek it to the position we need, and look at its [position(), buffer().end()] - just the most basic usage of ReadBuffer. For the seeks to work, we need that either (a) seeks are allowed (as the TODO [3] says), or (b) the buffer's position is already correct, so the seek is no-op. We can further assert (b) directly if we want to, but the rest CachedOnDiskReadBufferFromFile doesn't seem to care about (a) vs (b), it just needs `implementation_buffer->getPosition()` to be correct.

(I'm probably misunderstanding more things here, and it probably won't work, please bear with me :) )

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 11]
```diff
@@ -1192,12 +1185,10 @@ String CachedOnDiskReadBufferFromFile::getInfoForLog()
 
     return fmt::format(
         "Buffer path: {}, hash key: {}, file_offset_of_buffer_end: {}, "
-        "internal buffer remaining read range: {}, "
```
[the remark this answers is not in the document]
Comment (al13n321) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:1195]: Ok, I'll print `implementation_buffer->getFileOffsetOfBufferEnd()`, `read_until_position`, and `current_file_segment_it->range().right`. Or should I bring back the whole `getRemainingReadRange()`? It seems redundant, there are already `getPosition()` and `getFileOffsetOfBufferEnd()` (aren't these two already redundant? :( ); there's no `getReadUntilPosition()`, but the caller can usually keep track of it? (Is it just me or is the set of virtual methods in ReadBuffer and SeekableReadBuffer a mess?)

**Comment by al13n321** [comment]
Undid most of the nonsense refactoring in CachedOnDiskReadBufferFromFile. Switched from making it accept pre-existing reader in any state to making sure that pre-existing reader is always in the correct state.

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (kssenii) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:836]: > to worry less about missing a matching swap() later (on exception or early return)

btw there is a `swapHelper` struct somewhere in the code for this purpose

> wouldn't that leave the CachedOnDiskReadBufferFromFile in invalid state, with internal_buffer empty?

I do not think so, it will mean that we will lose the buffer which we passed through to be later passed to impl, but if we get an exception - we will not need this buffer anymore

>  I guess that's fine because we don't expect the CachedOnDiskReadBufferFromFile to be used after its nextImpl() throws?

yes

> Would it still be better to implementation_buffer->set(...)

so it is also an option, but not sure if it will make the difference

> This doesn't seem to hold for FileSegment::remote_file_reader - it can be reused by a later CachedOnDiskReadBufferFromFile even if the ReadBuffer threw an exception into an earlier CachedOnDiskReadBufferFromFile; right?

hm, yes, but as I understand it can result in a problem only in case we successfully read via this reader and then failed to write the result into file segment, but this write fails (or if the command before write - reserve - fails) then file_segment's state is changed to the one, which will not allow any further usage of FileSegment::remote_reader. There is no operation in between these events, so I assume this case will be processed correctly.

**Review by Avogar — APPROVED** [review]
